qownnotes: add darwin support

totoroot opened the pull request:
###### Description of changes

- Add darwin support for QOwnNotes
- Refactor to make diffs cleaner going forward
- Add changelog and downloadPage meta attributes
- Upgrade from 22.12.3 to 23.1.0

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

totoroot commented:
Result of `nixpkgs-review pr 208752` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>qownnotes</li>
  </ul>
</details>

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,51 @@
-    ++ lib.optionals stdenv.isLinux [ qtwayland ];
+  linux = stdenv.mkDerivation {
+    inherit pname version src meta;
+    nativeBuildInputs = [ qmake qttools qtbase qtsvg qtdeclarative qtxmlpatterns qtwebsockets qtx11extras qtwayland ];
```
I guess most of them should go to `buildInputs`.

totoroot replied:
You're right. I took a look at some other Qt packages and fixed a few issues in 4ccfb32.
QOwnNotes was the first package I added to nixpkgs and I hadn't changed much since.

totoroot commented:
@ofborg eval

totoroot commented:
Please merge if there are no other issues with this, as there is already a new QOwnNotes release.

wegank commented:
I think it's possible to build QOwnNotes from source on aarch64-darwin, so I'm not sure if a binary version is better here.

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,69 @@
-{ mkDerivation, lib, stdenv, fetchurl
-, qmake, qttools, qtbase, qtsvg, qtdeclarative, qtxmlpatterns, qtwebsockets
-, qtx11extras, qtwayland
+{ lib
+, stdenv
+, fetchurl
+, qmake
+, qttools
+, qtbase
+, qtdeclarative
+, qtsvg
+, qtwayland
+, qtwebsockets
+, qtx11extras
+, qtxmlpatterns
+, makeWrapper
+, wrapQtAppsHook
```
The reformatting is unnecessary.

totoroot replied:
No, it isn't. According to a recent review by @fabaff, this is preferred going forward, as it leads to cleaner Git diffs, so it is apparent at first glance, which dependencies were added/removed on package updates.

wegank replied:
I would prefer the reformatting to be in a separate commit to make the diff of **this** single-commit PR cleaner. But anyway, I guess I'm okay with that.

totoroot replied:
I put it in a separate commit now

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,69 @@
-{ mkDerivation, lib, stdenv, fetchurl
-, qmake, qttools, qtbase, qtsvg, qtdeclarative, qtxmlpatterns, qtwebsockets
-, qtx11extras, qtwayland
+{ lib
```
And what's the favicon.png added in the commit??

totoroot replied:
Thanks, fixed. Was a mistake.

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,69 @@
+  ];
+
+  postInstall = ''
+    ln -s $out/bin/${appname} $out/bin/${pname}
```
Maybe we can just set `pname = QOwnNotes`? Not sure though.

totoroot replied:
I don't think so. I just added the symlink since I think it is just beneficial to have a lowercase binary in the path as well.
What do you think? A simple symlink shouldn't hurt, right?

wegank replied:
Then please symlink the darwin binary (to `$out/bin/${appname}`) before `ln -s $out/bin/${appname} $out/bin/${pname}`, so that we have two files in `$out/bin` under both Linux and macOS.

totoroot replied:
We can't have that. Since macOS uses a case-insensitive file system you cannot have `$out/bin/QOwnNotes` and `$out/bin/qownnotes`.

Linux users are typically used to lowercase binaries, so IMO it makes absolute sense to have a symlink, as I've been using QOwnNotes for quite some time and have often needed my ZSH autocomplete to tell me, that `qownnotes` should be `QOwnNotes` when trying to launch it from the terminal.

The proposed solution means that we have `$out/bin/QOwnNotes` and `$out/bin/qownnotes` on Linux and  the `$out/bin/qownnotes` wrapper on macOS to call it from the terminal.

The application itself would still be QOwnNotes on macOS, so in Spotlight, Alfred etc. it would show up like that.

totoroot commented:
Thanks for the information, that it build fine for you. However, I don't get why we would need another PR.
I also refactored a bit and added some bits and pieces that would be nice to have, so please let's stick to this PR.

totoroot commented:
@wegank Please recheck on darwin, as I don't have access to darwin machines until next week.

wegank commented:
> Thanks for the information, that it build fine for you. However, I don't get why we would need another PR.
> I also refactored a bit and added some bits and pieces that would be nice to have, so please let's stick to this PR.

The PR was about adding `qownnotes-bin`, whereas it is not now. I think the description of this PR should be modified.

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,69 @@
-  nativeBuildInputs = [ qmake qttools ];
+  dontWrapQtApps = true;
 
-  buildInputs = [ qtbase qtsvg qtdeclarative qtxmlpatterns qtwebsockets qtx11extras ]
-    ++ lib.optionals stdenv.isLinux [ qtwayland ];
+  preFixup = ''
+      wrapQtApp "$out/bin/${appname}"
+  '';
```
Is this a negative effect of the additional symbolic link?

totoroot replied:
@wegank No, this is needed anyways.
Does it build on darwin?

wegank replied:
I don't get why this is needed. QOwnNotes is clearly wrapped without `dontWrapQtApps = true;` and `preFixup`.

```
$ file result/bin/QOwnNotes 
result/bin/QOwnNotes: a /nix/store/m45cjiwspshg01wrk2pln2qndasky23r-bash-5.1-p16/bin/bash -e script, ASCII text executable
```

Also, after removing qtwayland on darwin, the build still fails with

```
/nix/store/dljzajdq589id89i0a7ypdl9bhiicwg0-hook/nix-support/setup-hook: line 115: /nix/store/9x201hgq41mmmbxnvj6qgvkkws84gg8c-qownnotes-22.12.3/bin/qownnotes: Too many levels of symbolic links
```

totoroot replied:
Fixed that. It needs wrapQtAppsHook on darwin as well.

wegank commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,69 @@
+    qtbase
+    qtdeclarative
+    qtsvg
+    qtwayland
```
This does not evaluate on darwin.

totoroot replied:
Fixed

totoroot commented:
@wegank You're right. I changed the description.

totoroot commented:
> > Thanks for the information, that it build fine for you. However, I don't get why we would need another PR.
> > I also refactored a bit and added some bits and pieces that would be nice to have, so please let's stick to this PR.
> 
> The PR was about adding `qownnotes-bin`, whereas it is not now. I think the description of this PR should be modified.

@wegank You're right. I changed the description.

wegank commented:
@totoroot While according to ofborg, the darwin support is not added...

totoroot committed:
qownnotes: add darwin support

totoroot committed:
qownnotes: refactor

for cleaner diffs going forward

totoroot committed:
qownnotes: add meta attributes

totoroot committed:
qownnotes: 22.12.3 -> 23.1.0

wegank reviewed (approved):
LGTM. Tested on aarch64-darwin.

<img width="640" alt="Capture d’écran 2023-01-08 à 20 37 57" src="https://user-images.githubusercontent.com/9713184/211215431-2ba37e5f-ca6c-43bc-8a3f-cb2124f4713d.png">

totoroot commented:
Result of `nixpkgs-review pr 208752` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>qownnotes</li>
  </ul>
</details>

totoroot commented:
Result of `nixpkgs-review pr 208752` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>qownnotes</li>
  </ul>
</details>

totoroot commented:
> LGTM. Tested on aarch64-darwin.
> 
> <img alt="Capture d’écran 2023-01-08 à 20 37 57" width="640" src="https://user-images.githubusercontent.com/9713184/211215431-2ba37e5f-ca6c-43bc-8a3f-cb2124f4713d.png">

@wegank Thanks for your help

SuperSandro2000 commented on pkgs/applications/office/qownnotes/default.nix:
```diff
@@ -1,29 +1,66 @@
+    qtxmlpatterns
+  ] ++ lib.optionals stdenv.isLinux [ qtwayland ];
+
+  postInstall =
+  # Create a lowercase symlink for Linux
+  lib.optionalString stdenv.isLinux ''
+    ln -s $out/bin/${appname} $out/bin/${pname}
```
```suggestion
  # Create a lowercase symlink for Linux
  postInstall = lib.optionalString stdenv.isLinux ''
    ln -s $out/bin/${appname} $out/bin/${pname}
```
